paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -123,31 +124,112 @@ def describe_instances(instance_ids, region=None, instance_filters=None):
 
 
 @register_autoscaling_component('aws_spot_fleet_request', CLUSTER_METRICS_PROVIDER_KEY)
-def spotfleet_metrics_provider(spotfleet_request_id, resource, pool_settings):
-    mesos_state = get_mesos_master().state
+def spotfleet_metrics_provider(spotfleet_request_id, resource, pool_settings, config_folder, dry_run=False):
     sfr = get_sfr(spotfleet_request_id, region=resource['region'])
-    if not sfr or not sfr['SpotFleetRequestState'] == 'active':
-        log.error("Ignoring SFR {0} that does not exist or is not active.".format(spotfleet_request_id))
+    if not sfr or sfr['SpotFleetRequestState'] == 'cancelled':
+        log.error("SFR not found, removing config file.".format(spotfleet_request_id))
+        cleanup_cancelled_sfr_config(spotfleet_request_id, config_folder, dry_run=dry_run)
         return 0, 0
-    sfr['ActiveInstances'] = get_spot_fleet_instances(spotfleet_request_id, region=resource['region'])
-    resource['sfr'] = sfr
-    desired_instances = len(sfr['ActiveInstances'])
-    instance_ips = get_sfr_instance_ips(sfr, region=resource['region'])
+    elif sfr['SpotFleetRequestState'] in ['cancelled_running', 'active']:
+        sfr['ActiveInstances'] = get_spot_fleet_instances(spotfleet_request_id, region=resource['region'])
+        resource['sfr'] = sfr
+        desired_instances = len(sfr['ActiveInstances'])
+        mesos_state = get_mesos_master().state
+        slaves = get_sfr_slaves(resource, mesos_state)
+        error = get_mesos_utilization_error(spotfleet_request_id,
+                                            resource=resource,
+                                            pool_settings=pool_settings,
+                                            slaves=slaves,
+                                            mesos_state=mesos_state,
+                                            desired_instances=desired_instances)
+    elif sfr['SpotFleetRequestState'] in ['submitted', 'modifying', 'cancelled_terminating']:
+        log.warning("Not scaling an SFR in state: {0} so {1}, skipping...".format(sfr['SpotFleetRequestState'],
+                                                                                  spotfleet_request_id))
+        return 0, 0
+    else:
+        log.error("Unexpected SFR state: {0} for {1}".format(sfr['SpotFleetRequestState'],
+                                                             spotfleet_request_id))
+        raise ClusterAutoscalingError
+    if is_aws_launching_sfr_instances(sfr) and sfr['SpotFleetRequestState'] == 'active':
+        log.warning("AWS hasn't reached the TargetCapacity that is currently set. We won't make any "
+                    "changes this time as we should wait for AWS to launch more instances first.")
+        return 0, 0
+    current, target = get_spot_fleet_delta(resource, error)
+    if sfr['SpotFleetRequestState'] == 'cancelled_running':
+        resource['min_capacity'] = 0
```

Comment:
I think the reason I didn't do this is that we get a more gradual removal by removing the min capacity and setting the error to -1. I might change this in the future though as it would read clearer just to set both to 0.